acceptance: TestDockerNodeJS is flaky

I've seen this fail a few times on bors, e.g. [here](https://teamcity.cockroachdb.com/viewLog.html?buildId=10336751&buildTypeId=Cockroach_BazelEssentialCi).


Jira issue: CRDB-28383

in case we lose the artifacts:
```
  1) failure smoke test
       causes the docker build to fail on a test failure:
     AssertionError [ERR_ASSERTION]: undefined undefined undefined
      at Context.it (base-test.js:21:14)
      at callFn (/usr/lib/node/mocha/lib/runnable.js:354:21)
      at Test.Runnable.run (/usr/lib/node/mocha/lib/runnable.js:346:7)
      at Runner.runTest (/usr/lib/node/mocha/lib/runner.js:442:10)
      at /usr/lib/node/mocha/lib/runner.js:560:12
      at next (/usr/lib/node/mocha/lib/runner.js:356:14)
      at /usr/lib/node/mocha/lib/runner.js:366:7
      at next (/usr/lib/node/mocha/lib/runner.js:290:14)
      at Immediate.<anonymous> (/usr/lib/node/mocha/lib/runner.js:334:5)
  2) sequelize
       can create a model with an inverted index:
     Uncaught Error: Connection terminated unexpectedly
      at Connection.con.once (/usr/lib/node/pg/lib/client.js:179:21)
      at TLSSocket.<anonymous> (/usr/lib/node/pg/lib/connection.js:123:10)
      at endReadableNT (_stream_readable.js:1064:12)
      at _combinedTickCallback (internal/process/next_tick.js:138:11)
      at process._tickCallback (internal/process/next_tick.js:180:9)
(node:23) UnhandledPromiseRejectionWarning: Error: This socket has been ended by the other party
    at TLSSocket.writeAfterFIN [as write] (net.js:364:12)
    at Connection.end (/usr/lib/node/pg/lib/connection.js:313:22)
    at global.Promise (/usr/lib/node/pg/lib/client.js:410:23)
    at new Promise (<anonymous>)
    at Client.end (/usr/lib/node/pg/lib/client.js:409:12)
    at Context.after (/mnt/data/node/client.js:51:10)
    at callFn (/usr/lib/node/mocha/lib/runnable.js:354:21)
    at Hook.Runnable.run (/usr/lib/node/mocha/lib/runnable.js:346:7)
    at next (/usr/lib/node/mocha/lib/runner.js:304:10)
    at Immediate._onImmediate (/usr/lib/node/mocha/lib/runner.js:334:5)
    at runCallback (timers.js:794:20)
    at tryOnImmediate (timers.js:752:5)
    at processImmediate [as _immediateCallback] (timers.js:729:5)
(node:23) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
(node:23) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
Unhandled rejection TypeError: Cannot read property 'currentRetry' of undefined
    at /usr/lib/node/mocha/lib/runner.js:563:28
    at done (/usr/lib/node/mocha/lib/runnable.js:301:5)
    at /usr/lib/node/mocha/lib/runnable.js:365:11
    at tryCatcher (/usr/lib/node/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/usr/lib/node/bluebird/js/release/promise.js:512:31)
    at Promise._settlePromise (/usr/lib/node/bluebird/js/release/promise.js:569:18)
    at Promise._settlePromise0 (/usr/lib/node/bluebird/js/release/promise.js:614:10)
    at Promise._settlePromises (/usr/lib/node/bluebird/js/release/promise.js:689:18)
    at Async._drainQueue (/usr/lib/node/bluebird/js/release/async.js:133:16)
    at Async._drainQueues (/usr/lib/node/bluebird/js/release/async.js:143:10)
    at Immediate.Async.drainQueues (/usr/lib/node/bluebird/js/release/async.js:17:14)
    at runCallback (timers.js:794:20)
    at tryOnImmediate (timers.js:752:5)
    at processImmediate [as _immediateCallback] (timers.js:729:5)
non-zero exit code: 2
*
* INFO: command left-over files in /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/TestDockerNodeJS/runMode=docker#01/d9d385ad
*
    dockercluster.go:717: unexpected extra event &{0 die} (after [])
    --- FAIL: TestDockerNodeJS/runMode=docker#01 (13.41s)
   
```

--- a/pkg/server/diagnostics/BUILD.bazel
+++ b/pkg/server/diagnostics/BUILD.bazel
@@ -55,12 +55,13 @@ go_test(
     size = "medium",
     srcs = [
         "main_test.go",
+        "reporter_test.go",
         "update_checker_test.go",
     ],
     args = ["-test.timeout=295s"],
+    embed = [":diagnostics"],
     tags = ["no-remote-exec"],
     deps = [
-        ":diagnostics",
         "//pkg/base",
         "//pkg/build",
         "//pkg/security/securityassets",
@@ -71,6 +72,7 @@ go_test(
         "//pkg/util/cloudinfo",
         "//pkg/util/leaktest",
         "//pkg/util/log",
+        "@com_github_mitchellh_reflectwalk//:reflectwalk",
         "@com_github_stretchr_testify//require",
     ],
 )

--- a/pkg/server/diagnostics/reporter.go
+++ b/pkg/server/diagnostics/reporter.go
@@ -320,6 +320,10 @@ func (r *Reporter) populateSQLInfo(uptime int64, sql *diagnosticspb.SQLInstanceI
 	sql.Uptime = uptime
 }
 
+// collectSchemaInfo is the "old" way of collecting schema information, and it
+// redacted all `*string` type fields in the table descriptors but not `string`
+// type fields. Check out `schematelemetry` package for a better data source for
+// collecting redacted schema information.
 func (r *Reporter) collectSchemaInfo(ctx context.Context) ([]descpb.TableDescriptor, error) {
 	startKey := keys.MakeSQLCodec(r.TenantID).TablePrefix(keys.DescriptorTableID)
 	endKey := startKey.PrefixEnd()
@@ -442,7 +446,7 @@ func anonymizeZoneConfig(dst *zonepb.ZoneConfig, src zonepb.ZoneConfig, secret s
 type stringRedactor struct{}
 
 func (stringRedactor) Primitive(v reflect.Value) error {
-	if v.Kind() == reflect.String && v.String() != "" {
+	if v.Kind() == reflect.String && v.String() != "" && v.CanSet() {
 		v.Set(reflect.ValueOf("_").Convert(v.Type()))
 	}
 	return nil

--- a/pkg/server/diagnostics/reporter_test.go
+++ b/pkg/server/diagnostics/reporter_test.go
@@ -1,0 +1,49 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package diagnostics
+
+import (
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/mitchellh/reflectwalk"
+	"github.com/stretchr/testify/require"
+)
+
+// TestStringRedactor_Primitive tests that fields of type `*string` will be
+// correctly redacted to "_", and all other field types ( including `string`)
+// will be unchanged.
+func TestStringRedactor_Primitive(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	type Foo struct {
+		A string
+		B *string
+		C map[string]string
+	}
+
+	string1 := "string 1"
+	string2 := "string 2"
+	foo := Foo{
+		A: string1,
+		B: &string2,
+		C: map[string]string{"3": "string 3"},
+	}
+
+	require.NoError(t, reflectwalk.Walk(foo, stringRedactor{}))
+	require.Equal(t, "string 1", string1)
+	require.Equal(t, "string 1", foo.A)
+	require.Equal(t, "_", string2)
+	require.Equal(t, "_", *foo.B)
+	require.Equal(t, "string 3", foo.C["3"])
+}

